File: web/src/types/loan.ts
// Loan types - mirrors internal/domain/loan.go

import type { Branch } from './branch'
import type { Customer } from './customer'
import type { Item } from './item'

export type LoanStatus =
  | 'active'
  | 'paid'
  | 'overdue'
  | 'defaulted'
  | 'renewed'
  | 'confiscated'

export type PaymentPlanType = 'single' | 'minimum_payment' | 'installments'

export const LOAN_STATUSES: { value: LoanStatus; label: string; color: string }[] = [
  { value: 'active', label: 'Activo', color: 'green' },
  { value: 'paid', label: 'Pagado', color: 'blue' },
  { value: 'overdue', label: 'Vencido', color: 'red' },
  { value: 'defaulted', label: 'En Mora', color: 'red' },
  { value: 'renewed', label: 'Renovado', color: 'purple' },
  { value: 'confiscated', label: 'Confiscado', color: 'gray' },
]

export const PAYMENT_PLAN_TYPES: { value: PaymentPlanType; label: string; description: string }[] = [
  { value: 'single', label: 'Pago Único', description: 'Un solo pago al vencimiento' },
  { value: 'minimum_payment', label: 'Pago Mínimo', description: 'Pagos mínimos mensuales + pago final' },
  { value: 'installments', label: 'Cuotas', description: 'Pagos iguales divididos en cuotas' },
]

export interface Loan {
  id: number
  loan_number: string
  branch_id: number
  customer_id: number
  item_id: number

  // Amounts
  loan_amount: number
  interest_rate: number
  interest_amount: number
  principal_remaining: number
  interest_remaining: number
  total_amount: number
  amount_paid: number

  // Late fees
  late_fee_rate: number
  late_fee_amount: number

  // Dates
  start_date: string
  due_date: string
  paid_date?: string
  confiscated_date?: string

  // Payment plan
  payment_plan_type: PaymentPlanType
  loan_term_days: number
  requires_minimum_payment: boolean
  minimum_payment_amount?: number
  next_payment_due_date?: string
  grace_period_days: number

  // Installments
  number_of_installments?: number
  installment_amount?: number

  // Status
  status: LoanStatus
  days_overdue: number

  // Renewal
  renewed_from_id?: number
  renewal_count: number

  // Notes
  notes?: string

  // Audit
  created_by?: number
  updated_by?: number
  created_at: string
  updated_at: string
  deleted_at?: string

  // Relations
  branch?: Branch
  customer?: Customer
  item?: Item
}

export interface LoanInstallment {
  id: number
  loan_id: number
  installment_number: number
  due_date: string
  principal_amount: number
  interest_amount: number
  total_amount: number
  amount_paid: number
  is_paid: boolean
  paid_date?: string
  created_at: string
  updated_at: string
}

export interface CreateLoanInput {
  branch_id: number
  customer_id: number
  item_id: number
  loan_amount: number
  interest_rate?: number
  payment_plan_type: PaymentPlanType
  loan_term_days?: number
  grace_period_days?: number
  number_of_installments?: number
  notes?: string
}

export interface RenewLoanInput {
  new_term_days: number
  new_interest_rate?: number
  pay_interest?: boolean
}

export interface LoanListParams {
  page?: number
  per_page?: number
  order_by?: string
  order?: 'asc' | 'desc'
  search?: string
  branch_id?: number
  customer_id?: number
  item_id?: number
  status?: LoanStatus
  payment_plan_type?: PaymentPlanType
  is_overdue?: boolean
  date_from?: string
  date_to?: string
}

// Calculate remaining balance
export function calculateRemainingBalance(loan: Loan): number {
  return loan.principal_remaining + loan.interest_remaining + loan.late_fee_amount
}

File: internal/domain/loan.go
package domain

import (
	"time"
)

// LoanStatus represents the status of a loan
type LoanStatus string

const (
	LoanStatusActive      LoanStatus = "active"
	LoanStatusPaid        LoanStatus = "paid"
	LoanStatusOverdue     LoanStatus = "overdue"
	LoanStatusDefaulted   LoanStatus = "defaulted"
	LoanStatusRenewed     LoanStatus = "renewed"
	LoanStatusConfiscated LoanStatus = "confiscated"
)

// PaymentPlanType represents the type of payment plan
type PaymentPlanType string

const (
	PaymentPlanSingle         PaymentPlanType = "single"
	PaymentPlanMinimumPayment PaymentPlanType = "minimum_payment"
	PaymentPlanInstallments   PaymentPlanType = "installments"
)

// Loan represents a pawn loan
type Loan struct {
	ID         int64 `json:"id"`
	LoanNumber string `json:"loan_number"`
	BranchID   int64 `json:"branch_id"`
	CustomerID int64 `json:"customer_id"`
	ItemID     int64 `json:"item_id"`

	// Amounts
	LoanAmount         float64 `json:"loan_amount"`
	InterestRate       float64 `json:"interest_rate"`
	InterestAmount     float64 `json:"interest_amount"`
	PrincipalRemaining float64 `json:"principal_remaining"`
	InterestRemaining  float64 `json:"interest_remaining"`
	TotalAmount        float64 `json:"total_amount"`
	AmountPaid         float64 `json:"amount_paid"`

	// Late fees
	LateFeeRate   float64 `json:"late_fee_rate"`
	LateFeeAmount float64 `json:"late_fee_amount"`

	// Dates
	StartDate       Date       `json:"start_date"`
	DueDate         Date       `json:"due_date"`
	PaidDate        *time.Time `json:"paid_date,omitempty"`
	ConfiscatedDate *time.Time `json:"confiscated_date,omitempty"`

	// Payment plan
	PaymentPlanType        PaymentPlanType `json:"payment_plan_type"`
	LoanTermDays           int             `json:"loan_term_days"`
	RequiresMinimumPayment bool            `json:"requires_minimum_payment"`
	MinimumPaymentAmount   *float64        `json:"minimum_payment_amount,omitempty"`
	NextPaymentDueDate     *time.Time      `json:"next_payment_due_date,omitempty"`
	GracePeriodDays        int             `json:"grace_period_days"`

	// Installments
	NumberOfInstallments *int     `json:"number_of_installments,omitempty"`
	InstallmentAmount    *float64 `json:"installment_amount,omitempty"`

	// Status
	Status      LoanStatus `json:"status"`
	DaysOverdue int        `json:"days_overdue"`

	// Renewal info
	RenewedFromID *int64 `json:"renewed_from_id,omitempty"`
	RenewalCount  int    `json:"renewal_count"`

	// Notes
	Notes string `json:"notes,omitempty"`

	// Audit
	CreatedBy int64  `json:"created_by,omitempty"`
	UpdatedBy *int64 `json:"updated_by,omitempty"`

	// Timestamps
	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
	DeletedAt *time.Time `json:"deleted_at,omitempty"`

	// Relations
	Branch   *Branch   `json:"branch,omitempty"`
	Customer *Customer `json:"customer,omitempty"`
	Item     *Item     `json:"item,omitempty"`
}

// TableName returns the database table name
func (Loan) TableName() string {
	return "loans"
}

// RemainingBalance returns the total remaining balance
func (l *Loan) RemainingBalance() float64 {
	return l.PrincipalRemaining + l.InterestRemaining + l.LateFeeAmount
}

// IsOverdue checks if the loan is overdue
func (l *Loan) IsOverdue() bool {
	if l.Status != LoanStatusActive {
		return false
	}
	return time.Now().After(l.DueDate.Time)
}

// IsInGracePeriod checks if the loan is in grace period
func (l *Loan) IsInGracePeriod() bool {
	if !l.IsOverdue() {
		return false
	}
	gracePeriodEnd := l.DueDate.AddDate(0, 0, l.GracePeriodDays)
	return time.Now().Before(gracePeriodEnd)
}

// DaysUntilDue returns the number of days until due date
func (l *Loan) DaysUntilDue() int {
	days := int(time.Until(l.DueDate.Time).Hours() / 24)
	if days < 0 {
		return 0
	}
	return days
}

// CalculateDaysOverdue returns the number of days overdue
func (l *Loan) CalculateDaysOverdue() int {
	if !l.IsOverdue() {
		return 0
	}
	return int(time.Since(l.DueDate.Time).Hours() / 24)
}

// LoanInstallment represents an installment for a loan
type LoanInstallment struct {
	ID                int64     `json:"id"`
	LoanID            int64     `json:"loan_id"`
	InstallmentNumber int       `json:"installment_number"`
	DueDate           time.Time `json:"due_date"`
	PrincipalAmount   float64   `json:"principal_amount"`
	InterestAmount    float64   `json:"interest_amount"`
	TotalAmount       float64   `json:"total_amount"`
	AmountPaid        float64   `json:"amount_paid"`
	IsPaid            bool      `json:"is_paid"`
	PaidDate          *time.Time `json:"paid_date,omitempty"`
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
}

// TableName returns the database table name
func (LoanInstallment) TableName() string {
	return "loan_installments"
}

// RemainingAmount returns the remaining amount for this installment
func (li *LoanInstallment) RemainingAmount() float64 {
	return li.TotalAmount - li.AmountPaid
}
